feat(ip-restriction): support user-defined configuration message

Applenice opened the pull request:
### What this PR does / why we need it:
fix: https://github.com/apache/apisix/issues/4428

Added custom message to ip-restriction plugin and no problem in my local tests. It looks like I need to learn how to write test cases, right?

### Pre-submission checklist:

* [X] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [X] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

spacewander commented on apisix/plugins/ip-restriction.lua:
```diff
@@ -36,6 +36,10 @@ local schema = {
                     items = {anyOf = core.schema.ip_def},
                     minItems = 1
                 },
+                message = {
```
Better to extract this field to avoid repeating it

tokers replied:
And add some limitations for the max length?

Applenice replied:
I'll figure out how to modify it, thanks

spacewander replied:
See https://github.com/apache/apisix/blob/07bfe5b9c03ad361cb71541eb02fd661f1323213/apisix/schema_def.lua#L33

Applenice replied:
ok👌

spacewander commented on docs/en/latest/plugins/ip-restriction.md:
```diff
@@ -70,6 +70,20 @@ curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f13
+            "127.0.0.1",
+            "113.74.26.106/24"
+        ],
+        "message": "Do you want to do something bad?"
```
Need to update `Attributes`

Applenice replied:
Sorry, I didn't understand😥, what does this attribute refer to?

spacewander replied:
https://github.com/apache/apisix/blob/master/docs/en/latest/plugins/ip-restriction.md#attributes
Need to update the table.

Applenice replied:
Thanks, now I understand😊,previously ignored this place

spacewander commented on docs/en/latest/plugins/ip-restriction.md:
````diff
@@ -70,6 +70,20 @@ curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f13
 
+If you want to use a custom message, you can configure it in the plugin section.
+
+```shell
````
Should be json

spacewander commented on docs/en/latest/plugins/ip-restriction.md:
````diff
@@ -70,6 +70,20 @@ curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f13
 }'
 ```
 
+If you want to use a custom message, you can configure it in the plugin section.
````
Use a custom message for what?

Applenice replied:
Thanks for the review, I saw https://github.com/apache/apisix/issues/4428, which mentioned ip-restriction plugin support DIY message, was marked `good first issue`, so wanted to try to implement it.  I'll make the changes mentioned above later.

spacewander replied:
Err. Actually I mean we should write it more clear in the documentation, as the user won't have the same context as us.

Applenice replied:
I understand, I will write more clearly later

spacewander commented on apisix/plugins/ip-restriction.lua:
```diff
@@ -27,40 +27,29 @@ local lrucache  = core.lrucache.new({
+        message = {
+            type = "string",
+            minLength = 1,
+            maxLength = 64,
```
64 looks too short to me. What about 4096?

Applenice replied:
Would 4096 be too long?What about 1024?😃

spacewander commented on docs/en/latest/plugins/ip-restriction.md:
```diff
@@ -41,9 +41,10 @@ in CIDR notation like 10.10.10.0/24 can be used.
-One of `whitelist` or `blacklist` must be specified, and they can not work
-together.
+One of `whitelist` or `blacklist` must be specified, and they can not work together.
+The message supports user-defined configuration.
```
The message can be user-defined.

Applenice replied:
ok,later modify

spacewander commented:
@Applenice 
Need to add test in https://github.com/apache/apisix/blob/master/t/plugin/ip-restriction.t

Applenice commented:
> @Applenice
> Need to add test in https://github.com/apache/apisix/blob/master/t/plugin/ip-restriction.t

Okay😊

Applenice commented:
I changed the maxLength value from 64 to 1024, updated the documentation and added test cases, but there is a problem.
For test cases that are out of maxLength range, it is possible to pass the test using a fixed 1025 length string, but it should not be a graceful operation, how to call it with code generation? I searched some sources and didn't find an answer, so I didn't add a test case for when maxLength is exceeded.

spacewander commented:
You can use `string.rep`, like this one: https://github.com/apache/apisix/blob/d4c7fbbe831bc12c50ba63f00757f0eee14eae15/t/node/upstream-mtls.t#L100

tokers commented:
@Applenice Please take a look, the CI failed.

Applenice commented:
> You can use `string.rep`, like this one:
> 
> https://github.com/apache/apisix/blob/d4c7fbbe831bc12c50ba63f00757f0eee14eae15/t/node/upstream-mtls.t#L100

Thanks, I'll try it later

Applenice commented:
> @Applenice Please take a look, the CI failed.

Yes, I saw `Failed tests:  2, 5, 8, 14, 17, 72`, I observed it when testing locally with `prove -Itest-nginx/lib -r t/plugin/ip-restriction.t`, I'll check the reason later
